src/dryrun.ts

```diff
@@ -167,3 +171,293 @@ export async function createDryrun({
+}
+
+class DryrunTraceLine {
+  error: string = '';
+  line: number = 0;
+  pc: number = 0;
+  scratch: TealValue[] = [];
+  stack: DryrunStackValue[] = [];
+
+  constructor(line: DryrunTraceLineResponse) {
+    this.error = line.error === undefined ? '' : line.error;
+    this.line = line.line;
+    this.pc = line.pc;
+    this.scratch = line.scratch;
+    this.stack = line.stack.map(
+      (sv: StackValueResponse) => new DryrunStackValue(sv)
+    );
+  }
+}
+
+class DryrunTrace {
+  trace: DryrunTraceLine[] = [];
+  constructor(t: DryrunTraceLineResponse[]) {
+    if (t === undefined) return;
+    this.trace = t.map((line) => new DryrunTraceLine(line));
+  }
+}
+
+interface DryrunTransactionResultResponse {
+  disassembly: string[];
+  appCallMessages: string[] | undefined;
+  localDeltas: AccountStateDelta[] | undefined;
+  globalDelta: EvalDeltaKeyValue[] | undefined;
+  cost: number | undefined;
+  logicSigMessages: string[] | undefined;
+  logicSigDisassemly: string[] | undefined;
+  logs: string[] | undefined;
+  appCallTrace: DryrunTrace | undefined;
+  logicSigTrace: DryrunTrace | undefined;
+}
+
+interface StackPrinterConfig {
+  maxValueWidth: number | undefined;
+  topOfStackFirst: boolean | undefined;
+}
+
+function truncate(str: string, maxValueWidth: number): string {
+  if (str.length > maxValueWidth && maxValueWidth > 0) {
+    return `${str.slice(0, maxValueWidth)}...`;
+  }
+  return str;
+}
+
+function scratchToString(
+  prevScratch: TealValue[],
+  currScratch: TealValue[]
+): string {
+  if (currScratch.length === 0) return '';
+
+  let newScratchIdx = null;
+  for (let idx = 0; idx < currScratch.length; idx++) {
+    if (idx > prevScratch.length) {
+      newScratchIdx = idx;
+      continue;
+    }
+
+    if (JSON.stringify(prevScratch[idx]) !== JSON.stringify(currScratch[idx])) {
+      newScratchIdx = idx;
+    }
+  }
+
+  if (newScratchIdx == null) return '';
+
+  const newScratch = currScratch[newScratchIdx];
+  if (newScratch.bytes.length > 0) {
+    return `${newScratchIdx} = 0x${Buffer.from(
+      newScratch.bytes,
+      'base64'
+    ).toString('hex')}`;
+  }
+  return `${newScratchIdx} = ${newScratch.uint.toString()}`;
+}
+
+function stackToString(stack: DryrunStackValue[], reverse: boolean): string {
+  const svs = reverse ? stack.reverse() : stack;
+  return `[${svs
+    .map((sv: DryrunStackValue) => {
+      switch (sv.type) {
+        case 1:
+          return `0x${Buffer.from(sv.bytes, 'base64').toString('hex')}`;
+        case 2:
+          return `${sv.uint.toString()}`;
+        default:
+          return '';
+      }
+    })
+    .join(', ')}]`;
+}
+
+class DryrunTransactionResult {
+  disassembly: string[] = [];
+  appCallMessages: string[] | undefined = [];
+  localDeltas: AccountStateDelta[] | undefined = [];
+  globalDelta: EvalDeltaKeyValue[] | undefined = [];
+  cost: number | undefined = 0;
+  logicSigMessages: string[] | undefined = [];
+  logicSigDisassemly: string[] | undefined = [];
```

Comment:
noob question: is there a reason we distinguish the `logicSigDisassembly` (also check spelling) with the other `disassembly`? Or is it just because dryrun does this and we mirror it?

Reply:
Initially the dryrun endpoint only returned the last disassembly it saw, so in the case of an app call signed with logic you'd only get back the apps disassembly. In order to keep backwards compat I added lsig disassembly to the endpoint and dupe it in disassembly on algod. So they may be the same or different but more correct than before :)

Reply:
Got it thanks! LGTM - just need to change the remaining `logicSigDisassemly` to `logicSigDisassembly` I think and I'll approve